db/hints: Modernize manager

dawmd opened the pull request:
This PR is another step in refactoring the Hinted Handoff module. It aims at modernizing the code by moving to coroutines, using `std::ranges` instead of Boost's ones where possible, and uses other features coming with the new C++ standards.

It also tries to make the code clearer and get rid of confusing elements, e.g. using shared pointers where they shouldn't be used or marking methods as virtual even though nothing derives from the class. It also prevents `manager.hh` from giving direct access to internal structures (`hint_endpoint_manager` in this case).

Refs #15358

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4075/

avikivity commented on db/hints/manager.hh:
```diff
@@ -60,7 +60,8 @@ public:
 
-    ~directory_initializer();
+    ~directory_initializer() noexcept = default;
+
```
You may as well remove this line completely.

avikivity commented on db/hints/manager.cc:
```diff
@@ -439,15 +439,15 @@ bool manager::check_dc_for(endpoint_id ep) const noexcept {
 
     manager_logger.trace("on_leave_cluster: {} is removed/decommissioned", endpoint);
 
     // Future is waited on indirectly in `stop()` (via `_draining_eps_gate`).
```
Comment is now outdated.

In general, dropping a future (even with a comment) is bad practice, so this is a good change.

avikivity commented on db/hints/manager.cc:
```diff
@@ -241,7 +241,9 @@ void manager::allow_hints() {
-    boost::for_each(_ep_managers, [] (auto& pair) { pair.second.forbid_hints(); });
+    std::ranges::for_each(_ep_managers | boost::adaptors::map_values, [] (hint_endpoint_manager& ep_man) {
+        ep_man.forbid_hints();
+    });
```
Ditto, a regular for loop would be even simpler.

avikivity commented on db/hints/manager.cc:
```diff
@@ -286,36 +287,30 @@ future<> manager::wait_for_sync_point(abort_source& as, const sync_point::shard_
     }
```
> This commit changes the range we iterate through.
Instead of iterating over endpoint managers, we start
iterating over shard_rps. The rationale for that is
that usually, the size of shard_rps should be smaller
than the number of endpoint managers currently
managed by the hint shard manager. Also, if no endpoint
manager is found, we immediately return.

If this is pure waiting, then there are no negative side effects. However, if wait_until_...() also kicks something, then we may delay kicking it.

On the other hand, if it's pure, why does changing the loop nesting matter?

dawmd replied:
It is pure waiting indeed. The rationale for changing the range we loop over was to handle situations when we might have hundreds of endpoint managers, but there are just a few replay positions in the `shard_rps` that is passed to the function. Instead of creating hundreds of fibers, we would have only a few. Of course, the cost shouldn't be too big in either case, so it was more of a microoptimization rather than something having visible effect on performance.

However, it turns out that I must've been tired when refactoring the code because this specific commit changed the semantics of the function and I didn't notice it. Segments that were moved during hint rebalancing are treated differently; it more or less corresponds to the replay positions of those "foreign" segments being equal to minus infinity. Because of that, even if there is no replay position for a specific endpoint, we might still need to wait for the corresponding endpoint manager. As a result, iterating over all endpoint managers is inevitable.

Thank you for this comment. I'll go back to the previous semantics.

avikivity commented:
Good series, very easy to review. I left some minor comments.

dawmd commented:
V2:
* Rebased on top of the current `master`
* Removed a redundant explicit default destructor of `directory_initializer`
* Moved from arbitrary functors to `noncopyable_function` in `manager::with_file_update_mutex_for` and `hint_endpoint_manager::with_file_update_mutex`
* Commented why discarding futures is correct behavior in the right places
* Moved to regular for loops in `allow_hints`, `forbid_hints` and `forbid_hints_for_eps_with_pending_hints`
* Moved back to the previous semantics of `wait_for_sync_point` and commented it better
* Removed an unnecessary include (`#include <vector>`) I missed

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4107/
